File: src/server/src/velvet_dawn/mechanics/actions/action_modify.py
import enum
from typing import Union, Optional

import velvet_dawn.mechanics.selectors
from velvet_dawn import errors
from velvet_dawn.dao.models import TileInstance, UnitInstance
from velvet_dawn.dao.models.world_instance import WorldInstance
from velvet_dawn.mechanics.actions.action import Action
from velvet_dawn.mechanics.function_value import FunctionValue
from velvet_dawn.mechanics.selectors import Selector


""" The action modify class, responsible for modifying 
attributes of selector attached.
"""


class ActionModifierFunction(enum.Enum):
    """ Possible actions """
    SET = 0
    ADD = 1
    SUB = 2
    MUL = 3
    RESET = 4

    ADD_TAG = 5
    REMOVE_TAG = 6


""" The key:function map defined in the datapack dictionary """
KeyMap = {
    "set": ActionModifierFunction.SET,
    "add": ActionModifierFunction.ADD,
    "sub": ActionModifierFunction.SUB,
    "subtract": ActionModifierFunction.SUB,
    "mul": ActionModifierFunction.MUL,
    "multiply": ActionModifierFunction.MUL,
    "reset": ActionModifierFunction.RESET,
    "add-tag": ActionModifierFunction.ADD_TAG,
    "remove-tag": ActionModifierFunction.REMOVE_TAG,

    "set-selector": ActionModifierFunction.SET,
    "add-selector": ActionModifierFunction.SET,
    "sub-selector": ActionModifierFunction.SET,
    "subtract-selector": ActionModifierFunction.SET,
    "mul-selector": ActionModifierFunction.SET,
    "multiply-selector": ActionModifierFunction.SET,
}


NON_ATTRIBUTE_MODIFIERS = {ActionModifierFunction.ADD_TAG, ActionModifierFunction.REMOVE_TAG}


class ActionModify(Action):

    def __init__(self):
        super().__init__()

        self.selector: Optional[Selector] = None

        self.function: ActionModifierFunction = ActionModifierFunction.SET
        self.function_value = FunctionValue()
        self.is_value_selector = False

    @staticmethod
    def from_dict(id: str, data: dict):
        """ Parse the dict of tile/unit data to construct this action """
        if 'modify' not in data:
            raise errors.ValidationError("Modify functions must contain a 'modify' selector")

        # Check the number of functions is valid
        total_function_keys = sum([1 for key in data if key in KeyMap])
        if total_function_keys != 1:
            raise errors.ValidationError(f"Invalid modify action on {id}. Modify actions must contain one of {list(KeyMap)}. Problem data: {data}")

        # Construct the action and it's function
        action = ActionModify()
        action.selector = velvet_dawn.mechanics.selectors.get_selector(id, data['modify'])

        # Find the key used and update the function (we already verify there is a key above)
        function_key = None
        for key in data:
            if key in KeyMap:
                function_key = key
                action.function_value.parse(id, data[key])
                action.function = KeyMap[key]

        if action.function in NON_ATTRIBUTE_MODIFIERS:
            if action.selector.attribute is not None:
                raise errors.ValidationError(
                    f"Action '{function_key}' tags cannot be performed on a selector with attributes '{data['modify']}")
        else:
            if action.selector.attribute is None:
                raise errors.ValidationError(
                    f"Modify actions must ensure the selectors modify an attribute not '{data['modify']}'")

        return action

    def run(self, instance: Union[TileInstance, UnitInstance, WorldInstance]):
        """ Execute the attribute """
        if self.function == ActionModifierFunction.SET:
            self.selector.function_set(instance, self.function_value.value(instance))

        elif self.function == ActionModifierFunction.ADD:
            self.selector.function_add(instance, self.function_value.value(instance))

        elif self.function == ActionModifierFunction.MUL:
            self.selector.function_multiply(instance, self.function_value.value(instance))

        elif self.function == ActionModifierFunction.SUB:
            self.selector.function_subtract(instance, self.function_value.value(instance))

        elif self.function == ActionModifierFunction.RESET:
            self.selector.function_reset(instance, self.function_value.value(instance))

        elif self.function == ActionModifierFunction.ADD_TAG:
            self.selector.function_add_tag(instance, self.function_value.value(instance))

        elif self.function == ActionModifierFunction.REMOVE_TAG:
            self.selector.function_remove_tag(instance, self.function_value.value(instance))

        else:
            raise errors.ValidationError(f"Unknown action function type: '{self.function}'")

File: src/jserver/src/velvetdawn/mechanics/actions/ActionModify.java
package velvetdawn.mechanics.actions;

import velvetdawn.VelvetDawn;
import velvetdawn.mechanics.FunctionValue;
import velvetdawn.mechanics.selectors.Selector;
import velvetdawn.mechanics.selectors.Selectors;
import velvetdawn.models.instances.Instance;
import velvetdawn.utils.Json;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;

public class ActionModify extends Action {

    /* The action modify class, responsible for modifying
    attributes of selector attached.
    */

    public enum ActionModifierFunction {
        SET,
        ADD,
        SUB,
        MUL,
        RESET,
        ADD_TAG,
        REMOVE_TAG
    }

    /** The key:function map defined in the datapack dictionary */
    public static Map<String, ActionModifierFunction> KeyMap = Map.of(
            "set", ActionModifierFunction.SET,
            "add", ActionModifierFunction.ADD,
            "sub", ActionModifierFunction.SUB,
            "subtract", ActionModifierFunction.SUB,
            "mul", ActionModifierFunction.MUL,
            "multiply", ActionModifierFunction.MUL,
            "reset", ActionModifierFunction.RESET,
            "add-tag", ActionModifierFunction.ADD_TAG,
            "remove-tag", ActionModifierFunction.REMOVE_TAG
    );

    public static Set<ActionModifierFunction> NON_ATTRIBUTE_MODIFIERS = Set.of(
            ActionModifierFunction.ADD_TAG, ActionModifierFunction.REMOVE_TAG
    );

    private Selector selector;
    public ActionModifierFunction function = ActionModifierFunction.SET;
    private FunctionValue functionValue = null;

    /** Parse the dict of tile/unit data to construct this action */
    public static ActionModify fromJson(VelvetDawn velvetDawn, String parentId, Json data) throws Exception {
        if (!data.containsKey("modify")) {
            throw new Exception("Modify functions must contain a 'modify' selector");
        }

        // Check the number of functions is valid
        AtomicInteger totalFunctionKeys = new AtomicInteger();
        KeyMap.forEach((key, value) -> totalFunctionKeys.addAndGet(data.containsKey(key) ? 1 : 0));
        if (totalFunctionKeys.get() != 1) {
            // TODO Test this error formats correct
            throw new Exception(
                    String.format(
                            "Invalid modify action on %s. Modify actions must contain one of %s. Problem data: %s",
                            parentId,
                            data.keys(),
                            data.toString()
                    )
            );
        }

        // Construct the action and it's function
        ActionModify action = new ActionModify();
        action.selector = Selectors.get(velvetDawn, parentId,
                data.get("modify")
                        .validateInstanceIsString(String.format("Modify value on action in '%s' must be a string as a selector", parentId))
                        .value
        );


        // Find the key used and update the function (we already verify there is a key above)
        String functionKey = null;
        for (String key: data.keys()) {
            if (KeyMap.containsKey(key)) {
                functionKey = key;
                // TODO Test if not string
                action.functionValue = new FunctionValue(velvetDawn, parentId, data.get(key));
                action.function = KeyMap.get(key);
            }
        }

        // If a non-attribute modifier is used with an attribute or no attribute is
        // given to an attribute modifier then throw an error
        if (NON_ATTRIBUTE_MODIFIERS.contains(action.function)) {
            if (action.selector.attribute != null)
                throw new Exception(String.format(
                        "Action '%s' tags cannot be performed on a selector with attributes '%s'",
                        functionKey,
                        data.get("modify").validateInstanceIsString("Modify selector must be a string").value
                ));
        } else if (action.selector.attribute == null)
            throw new Exception(
                    String.format(
                        "Modify actions must ensure the selectors modify an attribute not '%s'",
                        data.get("modify").validateInstanceIsString("Modify selector must be a string").value
                    ));

        return action;
    }

    /** Execute the attribute */
    public void run(Instance instance) throws Exception {
        if (this.function == ActionModifierFunction.SET)
            this.selector.funcSet(instance, this.functionValue.value(instance));

        else if (this.function == ActionModifierFunction.ADD)
            this.selector.funcAdd(instance, this.functionValue.value(instance));

        else if (this.function == ActionModifierFunction.MUL)
            this.selector.funcMul(instance, this.functionValue.value(instance));

        else if (this.function == ActionModifierFunction.SUB)
            this.selector.funcSub(instance, this.functionValue.value(instance));

        else if (this.function == ActionModifierFunction.RESET)
            this.selector.funcReset(instance, this.functionValue.value(instance));

        else if (this.function == ActionModifierFunction.ADD_TAG)
            this.selector.funcAddTag(instance, this.functionValue.value(instance));

        else if (this.function == ActionModifierFunction.REMOVE_TAG)
            this.selector.funcRemoveTag(instance, this.functionValue.value(instance));

        else
            throw new Exception(String.format("Unknown action function type: '%s'", this.function));
    }
}
